composer/datasets/hf_streaming.py

```diff
@@ -0,0 +1,243 @@
+        dataset._ex_iterable.kwargs['filepaths'] = device_filepaths
+
+        # Subsample dataset if shard is being shared among devices
+        # NOTE: Mapping is executed in batched mode for better CPU utilization,
+        # but the returned dataset is still an iterable over text samples
+        if devices_per_shard > 1:
+            dataset = dataset.map(
+                partial(self._subsample, device_offset),
+                batched=True,
+                batch_size=devices_per_shard,
+            )
+        return dataset
+
+    def _tokenize(self, text_batch):
+        # Convert a text batch to a token batch
+        if self.group_method == "truncate":
+            truncation = True
+            padding = 'max_length'
+            max_length = self.max_seq_len
+        else:
+            truncation = False
+            padding = False
+            max_length = None
+        return self.tokenizer(text_batch["text"], truncation=truncation, padding=padding, max_length=max_length)
+
+    def _group_tokens(self, token_batch):
+        if self.group_method == "concat":
+            # Concatenate all tokens.
+            concat_tokens = {}
+            num_tokens = None
+            for k, v in token_batch.items():
+                concat_v = list(chain(*v))
+                concat_tokens[k] = concat_v
+                if num_tokens is None:
+                    num_tokens = len(concat_v)
+                else:
+                    assert num_tokens == len(concat_v), "Not all values in concat_tokens dict have same len()"
+            assert num_tokens is not None, "Failed to determine num_tokens."
+
+            # We drop the small remainder of tokens at the end of the batch,
+            # In the future we could support padding.
+            if num_tokens >= self.max_seq_len:
+                num_tokens = (num_tokens // self.max_seq_len) * self.max_seq_len
+
+            # Split into token samples of size max_seq_len.
+            result = {
+                k: [v[i:i + self.max_seq_len] for i in range(0, num_tokens, self.max_seq_len)
+                   ] for k, v in concat_tokens.items()
+            }
+            result["labels"] = result["input_ids"].copy()
+            return result
+        else:
+            raise ValueError(f"Unknown group_method: '{self.group_method}'")
+
+    def initialize_object(self, batch_size: int, dataloader_hparams: DataloaderHparams) -> DataSpec:
+        assert dataloader_hparams.num_workers == 1, "LM Streaming Dataloader only supports num_workers=1"
+
+        try:
+            import datasets
+            import transformers
+        except ImportError:
+            raise ImportError('HuggingFace transformers and datasets are not installed. '
+                              'Please install with `pip install mosaicml-composer[nlp]`')
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(self.tokenizer_name)  #type: ignore (thirdparty)
```

Comment:
What is the type error?

Reply:
Code leftover from previous implementation, removing the ignore.